diff -Naur tls_psk_state_machines.py tls_psk_state_machines.py
--- tls_psk_state_machines.py	2021-10-07 22:25:08.000000000 +0200
+++ tls_psk_state_machines.py	2021-10-20 14:42:56.495419000 +0200
@@ -111,8 +111,10 @@
         msg_len = int.from_bytes(
             curr_msg[curr_pos:curr_pos + tls_constants.MSG_LEN_LEN], 'big')
         if msg_type == tls_constants.APPLICATION_TYPE:
-            ptxt_msg = self.recv_hs_enc_connect.dec_packet(
+            type, ptxt_msg = self.recv_hs_enc_connect.dec_packet(
                 curr_msg[:msg_len])
+            if type != tls_constants.HANDSHAKE_TYPE:
+                raise InvalidMessageStructureError()
             msg_type = ptxt_msg[0]
             if msg_type == tls_constants.FINI_TYPE:
                 self.handshake.tls_13_process_finished(ptxt_msg)
diff -Naur tls_state_machines.py tls_state_machines.py
--- tls_state_machines.py	2021-10-07 22:25:08.000000000 +0200
+++ tls_state_machines.py	2021-10-20 14:42:41.487189000 +0200
@@ -111,8 +111,10 @@
         msg_len = int.from_bytes(
             curr_msg[curr_pos:curr_pos + tls_constants.MSG_LEN_LEN], 'big')
         if msg_type == tls_constants.APPLICATION_TYPE:
-            ptxt_msg = self.recv_hs_enc_connect.dec_packet(
+            type, ptxt_msg = self.recv_hs_enc_connect.dec_packet(
                 curr_msg[:msg_len])
+            if type != tls_constants.HANDSHAKE_TYPE:
+                raise InvalidMessageStructureError()
             msg_type = ptxt_msg[0]
             if msg_type == tls_constants.FINI_TYPE:
                 self.handshake.tls_13_process_finished(ptxt_msg)
